### PR TITLE
#10849 Liquid syntax support for CorrelateTask

### LINKED ISSUE #10849: Correlate Task should support Liquid AND javascript parsing with separate editors
The correlate task now uses the Monaco editor but only supports Script parsing.
We should also add the option the be able to parse Liquid by toggling the parser with a parser selector.
Should default to Script parsing to be backward compatible.

@hyzx86

### PR DESCRIPTION
**Pull request by lampersky**
fixes #10849

this is how it works:

https://user-images.githubusercontent.com/6403130/179355188-7f62d94a-532b-4e02-b18d-9a1e45772c2e.mp4

/cc @Skrypt

### CODE REVIEW

**Comment by hishamco** [comment]
Seems there's a breaking changes

**Comment by lampersky** [comment]
> Seems there's a breaking changes

@hishamco  no, there is not

here you have recipe from OC before that PR, and you can test it with this PR, and see that, this feature has backward compatibility

[Recipe.zip](https://github.com/OrchardCMS/OrchardCore/files/9125911/Recipe.zip)

File: src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs:16]: @lampersky this what I mean, just we need to add this on the module docs

Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs:16]: @hishamco what do you exactly want to add there?

image

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs:16]: No no, seems I was wrong, may be we need to add some docs [here](https://github.com/OrchardCMS/OrchardCore/blob/main/src/docs/releases/1.5.0.md)

File: src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs:37]: Let's use a string instead, so the syntax config is a bit more flexible. Most possibly never needed, but the general experience with such `bool IsOneOfTheThings` values is that they tend to get a third option with time too, and then it's messy to change.

That said... `IWorkflowExpressionEvaluator` is about a generic expression, and while its currently only (non-test) implementation is `LiquidWorkflowExpressionEvaluator`, this doesn't need to be the case. So maybe rather `IsExpressionSyntax` and change the rest of the UI wording too? Hmm, but then we can't have proper syntax highlighting.

Yeah, let's go with a string, and if we were to have multiple expression engines then we can revisit.

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:9]:
```suggestion
            @T["Choose between Liquid or JavaScript support."]
```

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:20]:
```suggestion
    <span id="javascript_hint" class="hint @(Model.IsLiquidSyntax ? "d-none" : "")">@T["The value to correlate workflow instances with. With JavaScript support."]</span>
```

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:55]: I'm not sure if related to this, but the textbox is tiny for me:

image

**Review by Piedone — CHANGES_REQUESTED** [review]
I sincerely apologize for us taking so much time here. I checked out the PR, going over old ones, and this would be quite useful.

How about other activities, see this comment? https://github.com/OrchardCMS/OrchardCore/issues/10849#issuecomment-990623411

Is triage still needed here, @Skrypt? And do you also want to review this?

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:55]: > I'm not sure if related to this, but the textbox is tiny for me:

It was implemented 2 years ago. I remember that when I tested it, the editor was bigger. You can see it on demo video attached.

Anyway, I will try to find some time to fix spotted issues on upcoming weekend. Thx @Piedone

**Comment by Skrypt** [comment]
@piedone maybe we should take a look more often to PR marked with "need triage" on thursday meetings. I've been trying to make @sebastienros to take a look at those by marking these with that tag. I think I had no comment to add on the PR and it just needed triage.

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:55]: Thank you!

**Comment by Piedone** [comment]
Sure, please bring them up. I think we only need to triage PRs on the meeting when there's something to discuss with the wider audience (this PR is like that, then?), i.e. you as a reviewer aren't sure.

Please keep to these @lampersky:

- Apply code suggestions directly so the reviewer doesn't have to eyeball the changes. These resolve themselves after applying them, that's fine.
- Don't resolve other conversations so it's easier to track for the reviewer. Then, the reviewer will resolve them.
- Feel free to mark conversations that you addressed to keep track of them with an emoji or otherwise, just don't resolve them.
- Please keep conversations happening in line comment in those convos, otherwise communication will be a mess. If you have trouble finding them, see [this video](https://github.com/OrchardCMS/OrchardCore/pull/14749#issuecomment-1917976028).
- Please click "Re-request review" in the top-right corner for each reviewer when you're ready for another round of review, so they know that you're done.

File: src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs]: done

File: src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs:49]:
```suggestion
                _ => throw new NotSupportedException($"The syntax {Syntax} isn't supported for CorrelateTask.") 
```

**Review by Piedone — APPROVED** [review]
@MikeAlhayek @sebastienros could you please check this during the next triage too?

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:56]: Is we know the exact values here, I don't think you'll need to lower case.

File: src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs:35]: For preexisting Tasks we need `"JavaScript"` here, so they'll continue to run without any user interaction.

File: src/OrchardCore.Modules/OrchardCore.Workflows/Drivers/CorrelateTaskDisplayDriver.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Drivers/CorrelateTaskDisplayDriver.cs:19]: A string makes it extensible, like `DisplayType` for shapes.

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:7]: I deliberately skipped it.

> The content of this attribute represents the value to be submitted with the form, should this option be selected. If this attribute is omitted, the value is taken from the text content of the option element.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/option#value

File: src/OrchardCore.Modules/OrchardCore.Workflows/Drivers/CorrelateTaskDisplayDriver.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Drivers/CorrelateTaskDisplayDriver.cs:19]: I don't think you need it extensible. The switch statement is evaluating specific values and that logic is not extensible.

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:7]: Yeas but if you provide a value, you can allow one to write the word JavaScript and Liquid in their own language. T["JavaScript"]

File: src/OrchardCore.Modules/OrchardCore.Workflows/Drivers/CorrelateTaskDisplayDriver.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Drivers/CorrelateTaskDisplayDriver.cs:19]: Yeah, indeed.

File: src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs:38]: If enum is used, then we don't need this

**Comment by Piedone** [comment]
When you're done addressing all feedback of a review, click ["Re-request review"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#re-requesting-a-review) in the top-right corner for each reviewer when you're ready for another round of review, so they know that you're done.

File: src/OrchardCore.Modules/OrchardCore.Workflows/Drivers/CorrelateTaskDisplayDriver.cs
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Drivers/CorrelateTaskDisplayDriver.cs:19]: changed to enums

File: src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs:14]:
```suggestion
        private readonly IWorkflowExpressionEvaluator _expressionEvaluator;

```

File: src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Activities/CorrelateTask.cs:17]:
```suggestion
        public CorrelateTask(
            IWorkflowScriptEvaluator scriptEvaluator,
```

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:53]: nameof() if faster than .ToString()
```suggestion
            if (syntax == '@nameof(WorkflowScriptSyntax.JavaScript)') {
```

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:8]:
```suggestion
        <option value="@nameof(WorkflowScriptSyntax.JavaScript)">@T["JavaScript"]</option>
```

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:9]:
```suggestion
        <option value="@nameof(WorkflowScriptSyntax.Liquid)">@T["Liquid"]</option>
```

File: src/OrchardCore/OrchardCore.Workflows.Abstractions/Models/WorkflowScriptSyntax.cs
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Workflows.Abstractions/Models/WorkflowScriptSyntax.cs:1]: Can we please use file-scope name spaces?

File: src/OrchardCore/OrchardCore.Workflows.Abstractions/Helpers/WorkflowScriptSyntaxExtensions.cs
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Workflows.Abstractions/Helpers/WorkflowScriptSyntaxExtensions.cs:4]:
```suggestion
namespace OrchardCore.Workflows.Helpers;

```

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:58]:
```suggestion
                editor.setOption('mode', '@nameof(WorkflowScriptSyntax.Liquid)');
```

Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:58]: Is this a valid value for the editor?

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:8]:
```suggestion
        <option value="@nameof(WorkflowScriptSyntax.JavaScript)">@T["JavaScript"]</option>
```

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:9]:
```suggestion
        <option value="@nameof(WorkflowScriptSyntax.Liquid)">@T["Liquid"]</option>
```

File: src/OrchardCore/OrchardCore.Workflows.Abstractions/Models/WorkflowScriptSyntax.cs
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Workflows.Abstractions/Models/WorkflowScriptSyntax.cs:6]:
```suggestion
        Liquid,
```

**Review by MikeAlhayek — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore.Workflows.Abstractions/Models/WorkflowScriptSyntax.cs
Comment (lampersky) [reply src/OrchardCore/OrchardCore.Workflows.Abstractions/Models/WorkflowScriptSyntax.cs:6]: this will be the first enum with such convention, I've just checked and all enums in the project are without comma after last item, so I will leave it as is

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (lampersky) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:58]: yes, it is

**Comment by Piedone** [comment]
@MikeAlhayek this is waiting for your review.

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/CorrelateTask.Fields.Edit.cshtml:19]:
```suggestion
    <span id="liquid_hint" class="hint @(Model.Syntax == WorkflowScriptSyntax.Liquid ? string.Empty : "d-none")">@T["The value to correlate workflow instances with. With Liquid support."]</span>
    <span id="javascript_hint" class="hint @(Model.Syntax == WorkflowScriptSyntax.JavaScript ? string.Empty : "d-none")">@T["The value to correlate workflow instances with. With JavaScript support."]</span>
```

**Review by MikeAlhayek — APPROVED** [review]

**Pull request merged**
